modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java

```diff
@@ -1515,121 +1534,122 @@ public KBArticle updateKBArticle(
-			}
-			else {
-				kbArticle.setStatus(WorkflowConstants.STATUS_DRAFT);
-			}
+			indexer.delete(oldKBArticle);
+		}
+		else {
+			kbArticle = oldKBArticle;
+		}
 
-			kbArticle.setTitle(title);
-			kbArticle.setContent(content);
-			kbArticle.setDescription(description);
-			kbArticle.setSections(
-				StringUtil.merge(KBSectionEscapeUtil.escapeSections(sections)));
-			kbArticle.setLatest(true);
-			kbArticle.setMain(false);
-			kbArticle.setSourceURL(sourceURL);
-			kbArticle.setDisplayDate(displayDate);
-			kbArticle.setExpirationDate(expirationDate);
-			kbArticle.setReviewDate(reviewDate);
-			kbArticle.setExpandoBridgeAttributes(serviceContext);
-
-			kbArticle = kbArticlePersistence.update(kbArticle);
-
-			if (oldKBArticle.isApproved() || oldKBArticle.isExpired()) {
-				oldKBArticle.setModifiedDate(oldKBArticle.getModifiedDate());
-				oldKBArticle.setLatest(false);
-
-				kbArticlePersistence.update(oldKBArticle);
-			}
+		if (oldKBArticle.isPending()) {
+			kbArticle.setStatus(WorkflowConstants.STATUS_PENDING);
+		}
+		else {
+			kbArticle.setStatus(WorkflowConstants.STATUS_DRAFT);
+		}
 
-			// Asset
+		kbArticle.setTitle(title);
+		kbArticle.setContent(content);
+		kbArticle.setDescription(description);
+		kbArticle.setSections(
+			StringUtil.merge(KBSectionEscapeUtil.escapeSections(sections)));
+		kbArticle.setLatest(true);
+		kbArticle.setMain(false);
+		kbArticle.setSourceURL(sourceURL);
+		kbArticle.setDisplayDate(displayDate);
+		kbArticle.setExpirationDate(expirationDate);
+		kbArticle.setReviewDate(reviewDate);
+		kbArticle.setExpandoBridgeAttributes(serviceContext);
 
-			updateKBArticleAsset(
-				userId, kbArticle, serviceContext.getAssetCategoryIds(),
-				serviceContext.getAssetTagNames(),
-				serviceContext.getAssetLinkEntryIds());
+		kbArticle = kbArticlePersistence.update(kbArticle);
 
-			// Attachments
+		if (oldKBArticle.isApproved() || oldKBArticle.isExpired()) {
+			oldKBArticle.setModifiedDate(oldKBArticle.getModifiedDate());
+			oldKBArticle.setLatest(false);
 
-			_addKBArticleAttachments(userId, kbArticle, selectedFileNames);
+			kbArticlePersistence.update(oldKBArticle);
+		}
 
-			_removeKBArticleAttachments(removeFileEntryIds);
+		// Asset
 
-			// Indexer
+		updateKBArticleAsset(
+			userId, kbArticle, serviceContext.getAssetCategoryIds(),
+			serviceContext.getAssetTagNames(),
+			serviceContext.getAssetLinkEntryIds());
 
-			_indexKBArticle(kbArticle);
+		// Attachments
 
-			// Workflow
+		_addKBArticleAttachments(userId, kbArticle, selectedFileNames);
 
-			_startWorkflowInstance(userId, kbArticle, serviceContext);
+		_removeKBArticleAttachments(removeFileEntryIds);
 
-			return kbArticle;
-		}
-		finally {
+		// Indexer
+
+		_indexKBArticle(kbArticle);
+
+		// Workflow
+
+		_startWorkflowInstance(userId, kbArticle, serviceContext);
+
+		if (autoLock && hasKBArticleLock(userId, resourcePrimKey)) {
 			unlockKBArticle(resourcePrimKey);
 		}
```

Comment:
BTW, why did you remove the `finally` ?

Reply:
I moved this to the other method, but you are right that it could make sense.

I'll restore it and forward the PR